commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/Mean.java

```diff
@@ -0,0 +1,195 @@
+     * from this mean.
+     *
+     * <p>See {@link Mean} for details on the computing algorithm.
+     *
+     * @param values Values.
+     * @return {@code Mean} instance.
+     */
+    public static Mean of(double... values) {
+        final Mean mean = Statistics.add(new StorelessMean(), values);
+
+        if (!Double.isFinite(mean.getAsDouble())) {
+            return mean;
+        }
+
+        final double xbar = mean.getAsDouble();
+        double correction = 0;
+
+        for (final double value : values) {
+            correction += value - xbar;
+        }
+
+        return StorelessMean.create(xbar + (correction / values.length), mean.getN());
+    }
+
+    /**
+     * Updates the state of the statistic to reflect the addition of {@code value}.
+     * @param value Value.
+     */
+    @Override
+    public abstract void accept(double value);
+
+    /**
+     * Gets the mean of all input values.
+     *
+     * <p>When no values have been added, the result is <code>NaN</code>.
+     *
+     * @return {@code Mean} of all values seen so far.
+     */
+    @Override
+    public abstract double getAsDouble();
+
+    /** {@inheritDoc} */
+    @Override
+    public abstract Mean combine(Mean other);
+
+    /**
+     * Returns the number of values that have been added.
+     * @return Number of values.
+     */
+    public abstract long getN();
+
+    /**
+     * {@code Mean} implementation that does not store the input value(s) processed so far.
+     */
+    private static class StorelessMean extends Mean {
+
+        /**
+         * External Moment used to compute the mean.
+         */
+        private FirstMoment firstMoment;
+
+        /**
+         * Creates a StorelessMean instance with an External Moment.
+         *
+         * @param m1 First moment.
+         * @param n Number of values.
+         */
+        private StorelessMean(final double m1, final long n) {
+            firstMoment = new FirstMoment(m1, n);
+        }
+
+        /**
+         * Create a Mean instance.
+         */
+        StorelessMean() {
+            firstMoment = new FirstMoment();
+        }
+
+
+        /**
+         * Creates a StorelessMean instance with an External Moment.
+         *
+         * @param m1 First moment.
+         * @param n Number of values.
+         * @return A StorelessMean instance.
+         */
+        static StorelessMean create(final double m1, final long n) {
+            return new StorelessMean(m1, n);
+        }
+
+        @Override
+        public void accept(double value) {
+            firstMoment.accept(value);
+        }
+
+        @Override
+        public double getAsDouble() {
+            return firstMoment.getAsDouble();
+        }
+
+        @Override
+        public Mean combine(Mean other) {
+            final StorelessMean that = (StorelessMean) other;
+            firstMoment = firstMoment.combine(that.firstMoment);
```

Comment:
Since we have created firstMomemt we know that the combine here will return a reference to the left-hand argument. As such we do not have to assign the result and the firstMoment can be final.